contracts/DataAvailability.sol

```diff
@@ -0,0 +1,212 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {AccessControlDefaultAdminRules} from '@openzeppelin/contracts/access/extensions/AccessControlDefaultAdminRules.sol';
+
+contract DataAvailability is AccessControlDefaultAdminRules {
+    /**
+     * @dev DataAvailability stores information about the objective function value for users
+     * across timestamps.
+     * 
+     * The contract creator and any permitted uploaders may upload user objective function data
+     * at a given timestamp. The objective function data that are uploaded represent the delta
+     * between the objective function value at the current timestamp and the value at the
+     * previous timestamp that data was uploaded. The data is not actually stored in the contract,
+     * but events are emitted for each upload, which can be used to reconstruct the full data history.
+     *
+     * Instead of storing all data in the contract, a rolling hash of the data is calculated and stored
+     * for each timestamp. When verifying the data, the submitted proof will contain a hash of the data
+     * computed in a trusted zkVM; if the hash contained in the proof matches the hash stored in the contract,
+     * the data is considered valid.
+
+     * The rolling hash is calculated by XORing the existing hash for a given timestamp with the hash of each
+     * user:value pair as they are received by the contract. Since XOR is both commutative and associative, this
+     * results in a hash that is invariant to the order of uploads. Since we cannot "remove" information about
```

Comment:
This is technically not true; we can "remove" a piece of data from the hash's history by just uploading it again, since A ^ A = 0, which would allow us to "correct" it later. This seems like a weird thing to do though, so I just decided to prevent that case here.